File: src/chain-observers/filters/content-metadata-filter.ts
import type { comment, TAccountName } from "@hiveio/wax";
import type { WorkerBee } from "../../bot";
import { ContentMetadataClassifier, ImpactedAccountClassifier, OperationClassifier } from "../classifiers";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { TFilterEvaluationContext } from "../factories/data-evaluation-context";
import { FilterBase } from "./filter-base";

// Base class for content filters (posts and comments)
export abstract class BlogContentMetadataFilter extends FilterBase {
  public constructor(
    worker: WorkerBee,
    private readonly reportAfterMsBeforePayout: number,
    accounts: TAccountName[],
    protected readonly isPost: boolean
  ) {
    super(worker);

    this.accounts = new Set(accounts);
  }

  protected readonly accounts: Set<TAccountName>;

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      ContentMetadataClassifier,
      OperationClassifier,
      ImpactedAccountClassifier
    ];
  }

  private queriedPosts: comment[] = [];

  public async match(data: TFilterEvaluationContext): Promise<boolean> {
    const { operationsPerType } = await data.get(OperationClassifier);

    const queryComments: comment[] = [...this.queriedPosts];
    this.queriedPosts = []; // Reset for next match

    if (operationsPerType.comment_operation)
      for(const { operation } of operationsPerType.comment_operation) {
        // Check if post/comment type matches what we're looking for
        const postIndicator = operation.parent_author === "";
        if (this.isPost !== postIndicator)
          continue;

        // Check author match
        if (!this.accounts.has(operation.author))
          continue;

        queryComments.push(operation);
      }

    if (queryComments.length > 0) {
      const comments = await data.query(ContentMetadataClassifier, {
        requestedData: queryComments,
        reportAfterMsBeforePayout: this.reportAfterMsBeforePayout
      });

      for(const operation of queryComments)
        if (!comments[operation.author] || !comments[operation.author][operation.permlink])
          this.queriedPosts.push(operation); // Add back to queried posts if not found

    }

    const { contentData } = await data.get(ContentMetadataClassifier);
    for(const _ in contentData)
      return true;

    return false;
  }
}

// Filter for comments (replies to posts or other comments)
export class CommentMetadataFilter extends BlogContentMetadataFilter {
  public constructor(
    worker: WorkerBee,
    reportAfterMsBeforePayout: number,
    accounts: TAccountName[]
  ) {
    super(worker, reportAfterMsBeforePayout, accounts, false);
  }
}

// Filter for posts (top-level content with empty parent_author)
export class PostMetadataFilter extends BlogContentMetadataFilter {
  public constructor(
    worker: WorkerBee,
    reportAfterMsBeforePayout: number,
    accounts: TAccountName[]
  ) {
    super(worker, reportAfterMsBeforePayout, accounts, true);
  }
}

File: python/workerbee/chain_observers/filters/content_metadata_filter.py
"""PostMetadataFilter / CommentMetadataFilter — content metadata filters.

Mirrors src/chain-observers/filters/content-metadata-filter.ts.

Uses ContentMetadataClassifier + OperationClassifier + ImpactedAccountClassifier.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .._ordered_set import OrderedSet
from .filter_base import FilterBase

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import DataEvaluationContext


class BlogContentMetadataFilter(FilterBase):
    """Base class for content metadata filters (posts and comments)."""

    def __init__(
        self,
        report_after_ms_before_payout: int,
        accounts: list[str],
        is_post: bool,
    ) -> None:
        super().__init__()
        self._report_after_ms_before_payout = report_after_ms_before_payout
        self._is_post = is_post
        self.accounts: OrderedSet[str] = OrderedSet(accounts)
        self._queried_posts: list[Any] = []

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        from ..classifiers.content_metadata_classifier import ContentMetadataClassifier
        from ..classifiers.impacted_account_classifier import ImpactedAccountClassifier
        from ..classifiers.operation_classifier import OperationClassifier

        return [ContentMetadataClassifier, OperationClassifier, ImpactedAccountClassifier]

    async def match(self, data: DataEvaluationContext) -> bool:
        from ..classifiers.content_metadata_classifier import ContentMetadataClassifier
        from ..classifiers.operation_classifier import OperationClassifier

        result = await data.get(OperationClassifier)
        operations_per_type = result["operations_per_type"]

        query_comments: list[Any] = [*self._queried_posts]
        self._queried_posts = []  # Reset for next match

        comment_ops = operations_per_type.get("comment_operation")
        if comment_ops:
            for entry in comment_ops:
                operation = entry["operation"]

                # Check if post/comment type matches what we're looking for
                post_indicator = operation["parent_author"] == ""
                if self._is_post != post_indicator:
                    continue

                # Check author match
                if operation["author"] not in self.accounts:
                    continue

                query_comments.append(operation)

        if len(query_comments) > 0:
            comments = await data.query(
                ContentMetadataClassifier,
                {
                    "requested_data": query_comments,
                    "report_after_ms_before_payout": self._report_after_ms_before_payout,
                },
            )

            for operation in query_comments:
                if not comments.get(operation["author"], {}).get(operation["permlink"]):
                    self._queried_posts.append(operation)  # Add back if not found

        content_result = await data.get(ContentMetadataClassifier)
        content_data = content_result["content_data"]
        return bool(content_data)


class CommentMetadataFilter(BlogContentMetadataFilter):
    """Filter for comment metadata (replies to posts or other comments)."""

    def __init__(self, report_after_ms_before_payout: int, accounts: list[str]) -> None:
        super().__init__(report_after_ms_before_payout, accounts, False)


class PostMetadataFilter(BlogContentMetadataFilter):
    """Filter for post metadata (top-level content with empty parent_author)."""

    def __init__(self, report_after_ms_before_payout: int, accounts: list[str]) -> None:
        super().__init__(report_after_ms_before_payout, accounts, True)
